File: julia/day8_part2_2019.jl

function main()
    imageData = strip(read("input.txt", String))
    width, height = 25, 6
    layerSize = width * height
    finalImage = fill('2', layerSize)

    for i in 1:layerSize:length(imageData)
        layer = imageData[i:min(i + layerSize - 1, end)]
        for (j, pixel) in enumerate(layer)
            finalImage[j] == '2' && (finalImage[j] = pixel)
        end
    end

    println("Decoded image:")
    for i in 0:(height - 1)
        for j in 0:(width - 1)
            pixel = finalImage[i * width + j + 1]
            print(pixel == '0' ? ' ' : '#')
        end
        println()
    end
end

main()

File: rust/day8_part2_2019.rs

use std::fs::File;
use std::io::{self, Read};

fn main() -> io::Result<()> {
    let mut file = File::open("input.txt")?;
    let mut image_data = String::new();
    file.read_to_string(&mut image_data)?;
    let image_data = image_data.trim();

    let width = 25;
    let height = 6;
    let layer_size = width * height;
    let mut final_image: Vec<u8> = vec![b'2'; layer_size];

    for layer in image_data.as_bytes().chunks(layer_size) {
        for (j, &pixel) in layer.iter().enumerate() {
            if final_image[j] == b'2' {
                final_image[j] = pixel;
            }
        }
    }

    println!("Decoded image:");
    for i in 0..height {
        for j in 0..width {
            let pixel = final_image[i * width + j];
            if pixel == b'0' {
                print!(" ");
            } else if pixel == b'1' {
                print!("#");
            }
        }
        println!();
    }
    Ok(())
}
